packages/framework/src/Model/Product/ProductVisibilityFacade.php

```diff
@@ -95,4 +101,76 @@ public function createAndRefreshProductVisibilitiesForPricingGroup(PricingGroup
     {
         $this->productVisibilityRepository->createAndRefreshProductVisibilitiesForPricingGroup($pricingGroup, $domainId);
     }
+
+    /**
+     * @param int[] $productIds
+     * @return bool[]
+     */
+    public function areProductsVisibleForDefaultPricingGroupOnSomeDomainIndexedByProductId(array $productIds): array
+    {
+        $productVisibilitiesIndexedByDomainId = [];
+
+        foreach ($this->domain->getAllIds() as $domainId) {
+            $productVisibilitiesIndexedByDomainId[$domainId] = $this->getProductsVisibilitiesByDomainIndexedByProductId(
+                $productIds,
+                $domainId,
+            );
+        }
+
+        $productVisibilitiesIndexedByProductId = array_fill_keys($productIds, false);
+
+        foreach ($productVisibilitiesIndexedByDomainId as $productVisibilities) {
+            foreach ($productVisibilities as $productId => $productVisibility) {
+                if ($productVisibilitiesIndexedByProductId[$productId] === false) {
+                    $productVisibilitiesIndexedByProductId[$productId] = $productVisibility;
+                }
+            }
+        }
+
+        return $productVisibilitiesIndexedByProductId;
+    }
+
+    /**
+     * @param int[] $productIds
+     * @return bool[]
+     */
+    public function areProductsVisibleForDefaultPricingGroupOnEachDomainIndexedByProductId(array $productIds): array
+    {
+        $productVisibilitiesIndexedByDomainId = [];
+
+        foreach ($this->domain->getAllIds() as $domainId) {
+            $productVisibilitiesIndexedByDomainId[$domainId] = $this->getProductsVisibilitiesByDomainIndexedByProductId(
+                $productIds,
+                $domainId,
+            );
+        }
+
+        $productVisibilitiesIndexedByProductId = array_fill_keys($productIds, true);
+
+        foreach ($productVisibilitiesIndexedByDomainId as $productVisibilities) {
+            foreach ($productVisibilities as $productId => $productVisibility) {
+                if ($productVisibilitiesIndexedByProductId[$productId]) {
+                    $productVisibilitiesIndexedByProductId[$productId] = $productVisibility;
+                }
+            }
+        }
+
+        return $productVisibilitiesIndexedByProductId;
+    }
+
```

Comment:
What about simplifying it like this? (Or something like this, it is just an idea and might be further refactored.

```
protected function getProductsVisibilityIndexedByProductId(array $productIds, bool $defaultVisibility): array
{
    $productVisibilitiesIndexedByProductId = array_fill_keys($productIds, $defaultVisibility);

    foreach ($this->domain->getAllIds() as $domainId) {
        $productVisibilities = $this->getProductsVisibilitiesByDomainIndexedByProductId($productIds, $domainId);

        foreach ($productVisibilities as $productId => $productVisibility) {
            if ($defaultVisibility) {
                $productVisibilitiesIndexedByProductId[$productId] = $productVisibilitiesIndexedByProductId[$productId] && $productVisibility;
            } else {
                $productVisibilitiesIndexedByProductId[$productId] = $productVisibilitiesIndexedByProductId[$productId] || $productVisibility;
            }
        }
    }

    return $productVisibilitiesIndexedByProductId;

public function areProductsVisibleForDefaultPricingGroupOnSomeDomainIndexedByProductId(array $productIds): array
    {
    return $this->getProductsVisibilityIndexedByProductId($productIds, false);
}

public function areProductsVisibleForDefaultPricingGroupOnEachDomainIndexedByProductId(array $productIds): array
    {
    return $this->getProductsVisibilityIndexedByProductId($productIds, true);
}
```

Reply:
btw, it looks like there could be simple tests for the two new public methods :slightly_smiling_face: